### ISSUE TITLE
BFS returns unexpected distance when multiple roots are given

### ISSUE THREAD

**Issue by Lucaweihs** [igraph/igraph#903]
I've come to this problem using the R package so I will give examples using R code as I'm unfamiliar with the C API.

Suppose we create a simple directed graph 1 -> 2 -> 3:

```
g = graph.edgelist(matrix(c(1,2,2,3), ncol=2, byrow=T))
```

Now suppose we want to find the distance of all nodes in g from the root nodes 1,2. Clearly these distances should be 0,0,1 (since 1,2 are both roots and 3 is a distance of 1 from 3). I expected to be able to do this with the code

```
graph.bfs(g, c(1,2), dist=T, neimode="out")$dist
```

but this gives distances 0,1,2. Looking at the igraph_bfs code in https://github.com/igraph/igraph/blob/master/src/visitors.c we see that this is because bfs is searched from each root one at a time rather than all being pushed onto the queue at once. Hence in the example above the distances are those from the first root and then the other roots are not searched since they were already visited.

One might argue that the above is not the intended use of roots but I think that this is confusing as starting bfs from a collection of nodes is a fairly natural operation.

**Comment by ntamas** [comment]
You are correct, `igraph_bfs` simply runs multiple BFS operations one at a time from each root. Since we cannot change this without breaking the API, the best we can do in the future is to add an `igraph_bfs_multi` function with a signature similar to `igraph_bfs` but with a behaviour that is identical to the one you described. In the R layer, we can probably do away with adding an optional keyword argument to `graph.bfs` that chooses between the two - but I'd rather leave that up to @gaborcsardi .

**Comment by gaborcsardi** [comment]
I don't mind creating a new function, but I am not sure when this would actually happen.

**Comment by szhorvat** [comment]
@vtraag This is assigned to you, and it _would_ be a breaking change. Are we going to do anything about this for 1.0 or postpone to 2.0? I'm not sure we have the time to do it for 1.0.
